sdk/Trace/ImmutableSpan.php

```diff
@@ -0,0 +1,154 @@
+<?php
+
+declare(strict_types=1);
+
+namespace OpenTelemetry\Sdk\Trace;
+
+use OpenTelemetry\Sdk\InstrumentationLibrary;
+use OpenTelemetry\Sdk\Resource\ResourceInfo;
+use OpenTelemetry\Trace as API;
+
+/**
+ * @psalm-immutable
+ */
+final class ImmutableSpan implements SpanData
+{
+    private Span $span;
+
+    /** @var non-empty-string */
+    private string $name;
+
+    private API\Links $links;
+    private API\Events $events;
+    private API\Attributes $attributes;
+    private int $totalAttributeCount;
+    private int $totalRecordedEvents;
+    private StatusData $status;
+    private int $endEpochNanos;
+    private bool $hasEnded;
+
+    /**
+     * @internal
+     * @psalm-internal OpenTelemetry\Sdk
+     *
+     * @param non-empty-string $name
+     */
+    public function __construct(
+        Span $span,
+        string $name,
+        API\Links $links,
+        API\Events $events,
+        API\Attributes $attributes,
+        int $totalAttributeCount,
+        int $totalRecordedEvents,
```

Comment:
> Is it possible to have `count($events) !== $totalRecordedEvents`?

Yea, in the case more events were added than the allowed limit. `$totalRecordedEvents` would be incremented but they wouldn't be actually added to the list of events. Using this method it's possible to calculate the amount of dropped events

But thanks for that link, maybe we could change the interface a bit to expose the dropped count based on `total - events.size` instead of just the total amount. That would probably be a bit more spec compliant.